gateway-js/src/supergraphManagers/IntrospectAndCompose/index.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  composeAndValidate,
+  compositionHasErrors,
+  ServiceDefinition,
+} from '@apollo/federation';
+import { Logger } from 'apollo-server-types';
+import { HeadersInit } from 'node-fetch';
+import resolvable from '@josephg/resolvable';
+import {
+  ServiceEndpointDefinition,
+  SupergraphSdlUpdateFunction,
+  SubgraphHealthCheckFunction,
+} from '../..';
+import {
+  getServiceDefinitionsFromRemoteEndpoint,
+  Service,
+} from './loadServicesFromRemoteEndpoint';
+import { SupergraphManager, SupergraphSdlHookOptions } from '../../config';
+
+export interface IntrospectAndComposeOptions {
+  subgraphs: ServiceEndpointDefinition[];
+  introspectionHeaders?:
+    | HeadersInit
+    | ((
+        service: ServiceEndpointDefinition,
+      ) => Promise<HeadersInit> | HeadersInit);
+  pollIntervalInMs?: number;
+  logger?: Logger;
+  subgraphHealthCheck?: boolean;
+}
+
+type State =
+  | { phase: 'initialized' }
+  | { phase: 'polling'; pollingPromise?: Promise<void> }
+  | { phase: 'stopped' };
+
+export class IntrospectAndCompose implements SupergraphManager {
+  private config: IntrospectAndComposeOptions;
+  private update?: SupergraphSdlUpdateFunction;
+  private healthCheck?: SubgraphHealthCheckFunction;
+  private subgraphs?: Service[];
+  private serviceSdlCache: Map<string, string> = new Map();
+  private timerRef: NodeJS.Timeout | null = null;
+  private state: State;
+
+  constructor(options: IntrospectAndComposeOptions) {
+    this.config = options;
+    this.state = { phase: 'initialized' };
+  }
+
+  public async initialize({ update, getDataSource, healthCheck }: SupergraphSdlHookOptions) {
+    this.update = update;
+
+    if (this.config.subgraphHealthCheck) {
+      this.healthCheck = healthCheck;
+    }
+
+    this.subgraphs = this.config.subgraphs.map((subgraph) => ({
+      ...subgraph,
+      dataSource: getDataSource(subgraph),
+    }));
+
+    let initialSupergraphSdl: string | null = null;
+    try {
+      initialSupergraphSdl = await this.updateSupergraphSdl();
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    // Start polling after we resolve the first supergraph
+    if (this.config.pollIntervalInMs) {
+      this.beginPolling();
+    }
+
+    return {
+      // on init, this supergraphSdl should never actually be `null`.
+      // `this.updateSupergraphSdl()` will only return null if the schema hasn't
+      // changed over the course of an _update_.
+      supergraphSdl: initialSupergraphSdl!,
+      cleanup: async () => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
```

Comment:
@glasser do you see a bug here? I might need a `stopping` state for this to be correct.

Reply:
I think this is OK. The point of the stopping phase in Gateway is to protect against multiple `stop` calls.  (Which is unlikely to happen since `stop` is generally called by ApolloServer.stop which *itself* has a similar pattern to protect against multiple stop calls due to multiple signals being received, but technically you can use Gateway's API directly.) This cleanup is never called directly by anything other than Gateway, so it shouldn't have to worry about it getting called multiple times because that's exactly what Gateway's own carefulness prevents. Right?

Reply:
I think the bug in question would be if `poll` could call itself again before we actually move to the `stopped` phase`. (between L83 and L85 could we potentially poll again?)